### PR TITLE
Use more generic way to mount vendor

### PR DESCRIPTION
**Pull request by TheScarastic**

### CODE REVIEW

**Comment by TheScarastic** [comment]
i agree, but i dont think there is better way to handle it

**Comment by auanasgheps** [comment]
Guys, I don't know if I am exact, but this can/should be addressed in the TWRP.
Actually this has been done for Xiaomi Mi5, an [updated TWRP](https://forum.xda-developers.com/showpost.php?p=70813439&postcount=2) has been released to use /cust as /vendor.

Probably Magisk should not mess around with this.

**Comment by TheScarastic** [comment]
uhh? that is not even related with full_treble enabled selinux is split and while trying to patch it init crashes

**Comment by Jerry981028** [comment]
Perhaps we should verify the phone model first.

**Comment by feherneoh** [comment]
Just an idea, but cannot magiskinit just have a look at the fstab from the original ramdisk, and set up the blockdevice for vendor based on that?

**Comment by TheScarastic** [comment]
@Dinolek pretty sure it wont work that way atleast for magisk since it doesn't looks in /dev/block instead it gets partition number from /sys/dev/block  https://github.com/topjohnwu/Magisk/blob/master/native/jni/core/magiskinit.c#L118

**Comment by Dinolek** [comment]
@TheScarastic /sys/dev/block are by by block numbers, not by dev name, at least in my device as magisk search.
I tried it and flashing magisk work with this symlink.

**Comment by feherneoh** [comment]
@Dinolek and what about booting?

**Comment by Dinolek** [comment]
@feherneoh @TheScarastic Yes, sorry, it doesn't boot with that symlink, I first thought that this is for flashing , not for booting

File: native/jni/core/magiskinit.c [hunk 1]
```diff
@@ -428,11 +428,24 @@ int main(int argc, char *argv[]) {
...
 		mkdir("/sys", 0755);
 		xmount("sysfs", "/sys", "sysfs", 0, NULL);
 
+		// Check if device uses early-mount
+		if (access("/sys/firmware/devicetree/base/firmware/android/fstab/vendor/dev", R_OK) == 0) {
+			fd = open("/sys/firmware/devicetree/base/firmware/android/fstab/vendor/dev", O_RDONLY | O_CLOEXEC);
+			vendorpath[read(fd, vendorpath, sizeof(vendorpath))] = '\0';
+			strcpy(vendor, strrchr(vendorpath, '/') + 1);
+			close(fd);
+		} else
```
Comment (TheComputerGuy96) [inline native/jni/core/magiskinit.c]: Where's the `}`?

File: native/jni/core/magiskinit.c [hunk 2]
```diff
@@ -428,11 +428,24 @@ int main(int argc, char *argv[]) {
...
+			strcpy(vendor, "vendor");
+
```
Comment (TheComputerGuy96) [inline native/jni/core/magiskinit.c]: And, where's the `{`?

Comment (dark-1) [reply native/jni/core/magiskinit.c]: My bad , I mean "Single line Code" for `else`

File: native/jni/core/magiskinit.c (hunk 1 above)
Comment (dark-1) [reply native/jni/core/magiskinit.c]: @TheScarastic ,
So that's the reason you said LOL ?????

**Comment by spidey** [comment]
@Jerry981028, unfortunately the binary you released is not working for me neither, on a Xiaomi Redmi Note 4X Snapdragon I'm using RedWolf recovery.

**Comment by Jerry981028** [comment]
@spidey Need detailed information, please open an issue in the my repo.

**Comment by topjohnwu** [comment]
I will find a way to configure a custom vendor partition without hardcoding for specific devices.
But before it's implemented, I'll still leave this PR open.

**Comment by TheScarastic** [comment]
@topjohnwu can you explain what do you mean by  "hardcoding* since current code dynamically looks up for vendor partition through device-tree early mount sysfs and mounts it, if no sysfs is present it fallback to traditional "vendor" partition

**Comment by sjayminsgithub** [comment]
Any update regarding this issue ??

**Comment by topjohnwu** [comment]
Yes :)

**Pull request closed without merging**